### ISSUE TITLE
How to set SIP Timer C?

### ISSUE THREAD

**Issue by calvet27** [davehorton/drachtio-server#135]
Dear Sir,

I checked current timers configuration,
But there no timer for SIP Timer C.
How could I set this via drachtio,.conf.xml?

Best Regards,
Jihyun Yu

**Comment by davehorton** [comment]
Good question.  I am not sure whether the sofia stack exposes a configuration setting for timer C.  I will check.

So you are referring to a scenario where drachtio server is acting as a sip proxy, correct?

**Comment by calvet27** [comment]
> Good question. I am not sure whether the sofia stack exposes a configuration setting for timer C. I will check.
> 
> So you are referring to a scenario where drachtio server is acting as a sip proxy, correct?

Yes. I'm using drachtio server as a sip proxy (gateway) for webrtc. 
Please let me know the results of setting Timer C.

**Comment by calvet27** [comment]
Is there any update?

**Comment by davehorton** [comment]
Yes, I have looked at the code and recalled that I actually set timer C (good news).  It is not currently exposed through a configuration variable, though that could be done.

However, this section of [RFC 3261](https://tools.ietf.org/html/rfc3261#section-16.8) seems pretty explicit about its duration:
```
11. Set timer C

         In order to handle the case where an INVITE request never
         generates a final response, the TU uses a timer which is called
         timer C.  Timer C MUST be set for each client transaction when
         an INVITE request is proxied.  The timer MUST be larger than 3
         minutes.
```
I presume you want to set it shorter than 3 minutes, which would violate the spec.  Why do you need a shorter duration - i.e. what is your use case?

**Comment by calvet27** [comment]
> Yes, I have looked at the code and recalled that I actually set timer C (good news). It is not currently exposed through a configuration variable, though that could be done.
> 
> However, this section of [RFC 3261](https://tools.ietf.org/html/rfc3261#section-16.8) seems pretty explicit about its duration:
> 
> ```
> 11. Set timer C
> 
>          In order to handle the case where an INVITE request never
>          generates a final response, the TU uses a timer which is called
>          timer C.  Timer C MUST be set for each client transaction when
>          an INVITE request is proxied.  The timer MUST be larger than 3
>          minutes.
> ```
> 
> I presume you want to set it shorter than 3 minutes, which would violate the spec. Why do you need a shorter duration - i.e. what is your use case?

Thanks for updating me.
Actually, I have to increase Timer C value up to 1 hour to meet our customer's requirement.
The shorter value is not my focus.
Please let me know the way to set Timer C value through the configuration.

Many Thanks.

**Comment by davehorton** [comment]
I will look into making it configurable

**Comment by calvet27** [comment]
Ok. Thanks. I will wait for your update.

**Comment by calvet27** [comment]
Any update here?

**Comment by davehorton** [comment]
I will try to test out this change this week, if possible.  I have not prioritized it highly because your use case seems unlikely to be of general interest.

**Comment by davehorton** [comment]
I have pushed a branch called 'timerc-proxy', can you test this there?

You can set a timer C value in the config, command-line or env vars.  This will *only* apply to proxy processing, not B2BUA.  The value is in units of seconds, and MUST be greater than 180.

For instance, to set timer c to 5 minutes:

via config file:
```
<sip>
   <proxy-timer-c>300</proxy-timer-c>

</sip>
```
via command line:
```
drachtio --proxy-timer-c 300
```
via env:
```
DRACHTIO_PROXY_TIMER_C=300 drachtio
```
Could you please test and let me know the results?

**Comment by calvet27** [comment]
Many thanks for your support.
But I use Drachtio sip server within Docker. 
Please let me know what the image I have to use for test 'timerc-proxy'.

**Comment by davehorton** [comment]
docker pull drachtio/drachtio-server:timerc-proxy

**Comment by calvet27** [comment]
Sorry to late reply.
I checked the results of setting timer C as 3600 secs with above docker image,
But it does not work. The call is canceled by default value 180 secs.
Please double checked it.

**Comment by calvet27** [comment]
Is there any update?

**Comment by calvet27** [comment]
Is there any update yet?

**Comment by davehorton** [comment]
sorry, will get a fix out today/tomorrow

**Comment by calvet27** [comment]
Thanks. When can I get the fixed docker image?

**Comment by calvet27** [comment]
When can I get the fixed docker image?

**Comment by davehorton** [comment]
I just pushed a commit and docker image is building now.  I tested with a timerC value of 300 seconds (5 mins) and it worked for me

**Comment by davehorton** [comment]
ok docker image is available, please retest and let me know the results.

**Comment by calvet27** [comment]
Thx. The tag name of the image is same?

**Comment by calvet27** [comment]
I tried to rest with the new image at same tag, timerc-proxy. But it is not worked with the timer C value 3600 seconds.

**Comment by davehorton** [comment]
Strange because it worked for me.  Can you share a complete drachtio log with loglevel = debug and sofia loglevel 9

**Comment by calvet27** [comment]
Ok. Please check the attached text file.

[drachtio_timerc_proxy_20201124.txt](https://github.com/davehorton/drachtio-server/files/5588312/drachtio_timerc_proxy_20201124.txt)

**Comment by davehorton** [comment]
Uh, I don't see any call being proxied in that log.  I see an outbound INVITE sent as a UAC, which is different

**Comment by calvet27** [comment]
Attached log describes that 1 UAC and 1 UAS with different call id respectively at same drachtio server.
So is there no way to resolve this in that case?

**Comment by davehorton** [comment]
timer C specifically and only applies to proxy transactions:

https://tools.ietf.org/html/rfc3261#page-265

And per your request, that is what I have done.  This means that it will be triggered when an app uses `srf.proxyRequest`

So now you are telling me you are not proxying the request ??

**Comment by calvet27** [comment]
I do not use srf.proxyRequest.
I use both srf.createUAC and srf.createUAS.

So I want to make the inbound call cannot be canceled by Drachtio itself.
And I want to set that ringing timer(i.e. no-answer wait timer) via Draachtio's configuration up to 1 hour.
Is it possible?
